lib/std/tar.zig

```diff
@@ -1,172 +1,1141 @@
+                        @field(hdr, f.name) = @field(new_hdr, f.name);
+                },
+                []const u8 => {
+                    // verify all []const u8 field defaults == ""
+                    const default = comptime fieldDefault(f);
+                    comptime assert(default.len == 0);
+                    if (@field(new_hdr, f.name).len != 0)
+                        @field(hdr, f.name) = @field(new_hdr, f.name);
+                },
+                // skip pax_recs which is only used for testing.
+                // NOTE: don't try to get fieldDefault() of this field which
+                // is void in non-testing modes. it will error with
+                // 'error: alignment must be >= 1'.
+                []const []const u8, void => assert(mem.eql(u8, f.name, "pax_recs")),
+                FormatSet => {
+                    const default = comptime fieldDefault(f);
+                    comptime assert(default.eql(FormatSet.initEmpty()));
+                    if (!@field(new_hdr, f.name).eql(default))
+                        @field(hdr, f.name) = @field(new_hdr, f.name);
+                },
+                else => @compileLog(comptime fmt.comptimePrint("todo {s}", .{@typeName(f.type)})),
+            }
+        }
     }
+};
 
-    pub fn fileType(header: Header) FileType {
-        const result = @intToEnum(FileType, header.bytes[156]);
-        return if (result == @intToEnum(FileType, 0)) .normal else result;
+pub fn unixTime(tv_sec: i64, tv_nsec: i64) i128 {
+    const result = @bitCast(i128, [_]i64{ tv_sec * time.ns_per_s, tv_nsec });
+    return result;
+}
+
+pub const block_len = 512;
+pub const Block = *[block_len]u8;
+
+const V7Header = extern struct {
+    bytes: Block,
+
+    /// File name - 0..100
+    inline fn fileName(h: V7Header) []const u8 {
+        return h.bytes[0..100];
+    }
+    /// File mode (octal) - 100..108
+    inline fn mode(h: V7Header) []const u8 {
+        return h.bytes[100..108];
+    }
+    /// Owner's numeric user ID (octal) - 108..116
+    inline fn uid(h: V7Header) []const u8 {
+        return h.bytes[108..116];
+    }
+    /// Group's numeric user ID (octal) - 116..124
+    inline fn gid(h: V7Header) []const u8 {
+        return h.bytes[116..124];
+    }
+    /// File size in bytes (octal) - 124..136
+    inline fn size(h: V7Header) []const u8 {
+        return h.bytes[124..136];
+    }
+    /// Last modification time in numeric Unix time format (octal) - 136..148
+    inline fn modTime(h: V7Header) []const u8 {
+        return h.bytes[136..148];
+    }
+    /// Checksum for header record - 148..156
+    inline fn checksum(h: V7Header) []const u8 {
+        return h.bytes[148..156];
+    }
+    /// Link indicator (file type) - 156..157
+    inline fn fileType(h: V7Header) FileType {
+        return @intToEnum(FileType, h.bytes[156]);
+    }
+    /// Name of linked file - 157..257
+    inline fn linkedFileName(h: V7Header) []const u8 {
+        return h.bytes[157..257];
     }
```

Comment:
Why do these functions poke at bytes? These can be extern structs of byte arrays. Same for `UstarHeader`, `StarHeader` and `GnuHeader`

Reply:
Sounds good. Will do. I was previously using extern structs with byte arrays but switched away for some reason that i can't remember now.  I think I was fumbling with a bitcast and settled on this.